Fix the package not loading if it has an accidental file left in the package root directory.

aleksmaus opened the pull request:
Fix the package not loading if it has an accidental file in the package root directory.
Log the error and load the remaining sibling directories.

Addresses https://github.com/elastic/package-registry/issues/672

<img width="896" alt="Screen Shot 2021-01-20 at 9 11 18 PM" src="https://user-images.githubusercontent.com/872351/105271095-e8787e00-5b64-11eb-91f2-5dd22f3c74b9.png">

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-673/detail/PR-673/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-673/detail/PR-673/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-673/detail/PR-673/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-673/detail/PR-673/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-registry_673.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #673 opened`



* * Start Time: 2021-01-21T02:18:07.366+0000


* Duration: 7 min 23 sec


* Commit: 8cd2384b569f3821daa8d064511b2da8f4a56eff



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 169  |
| Skipped      | 0 |
| Total        | 169   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->










<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ruflin commented on util/packages.go:
```diff
@@ -74,6 +74,12 @@ func getPackagePaths(allPaths []string) ([]string, error) {
+				// Fixes an annoying problem when the .DS_Store file is left behind and the package
+				// is not loading without any error information
+				log.Printf("error: unexpected file: %s", path)
+				return nil
```
I wonder if we should return an error instead. This is at first inconvenient for the developer working with the package because he has to fix it but it assures no "invalid" files are inside the package.

@mtojek @ycombinator Do we have any validation in place in `elastic-package check` that would detect these files?

mtojek replied:
I'm wondering how did @aleksmaus run into this issue? Have you been using the elastic-package to boot up the stack? Did you interact with OSX file browser?

@ruflin This directory is not part of a package (look at the path). It's a bug in package-registry. Speaking about validation, there is such one in place, but there are places in which it's not a bug but a feature, like e.g. a package root. It's hard to detect these files as the end-user may include additional READMEs, license notes, etc. 

For reference: https://github.com/elastic/package-spec/blob/master/versions/1/spec.yml#L8 

@aleksmaus I don't think we need to fail the entire application in case of noticing strange files. It will be more convenient to the end-user to skip this files. Could you try this piece of code?

```go
if info.IsDir() {
    log.Printf("%-20s\t%10s\t%s", dirs[0], dirs[1], path)
    foundPaths = append(foundPaths, path)
    return filepath.SkipDir
}
return nil
```

aleksmaus replied:
Mac OS finder creates these pesky .DS_Store files for Finder app. 
This change doesn't fail the whole app, it keeps loading the package versions. 
Before the change, stumbling on such file (not dir code path) was returning filepath.SkipDir which skips all the siblings directories. 
In my case the directory layout was 
```
.DS_Store
0.1.0
0.1.2
```

And the existing code was stumbling upon .DS_Store file first and since it was not directory the default code path was always returning filepath.SkipDir

From documentation of filepath.Walk: 
```
 If the function returns SkipDir when invoked on a non-directory file, Walk skips the remaining files in the containing directory.
 ```
 
 Sure can update the code  the sample you provided, it's pretty much the same what my code does.
 Would still like to keep the log statement pointing at the error during load.

ruflin reviewed (commented):
Can you add a changelog entry?

aleksmaus commented:
@mtojek @ruflin is this good to merge now?

ycombinator commented on util/packages.go:
```diff
@@ -74,8 +74,13 @@ func getPackagePaths(allPaths []string) ([]string, error) {
 			if info.IsDir() {
```
I would take this opportunity to make this check a bit stronger.

If the execution reaches this line, we know that we should be looking at a directory whose name is a version. So how about we not only check if `info.IsDir()` but also check that `dirs[1]` [is a semver](https://godoc.org/github.com/Masterminds/semver#StrictNewVersion)?  That way we handle not only the unexpected _file_ case but also the unexpected _folder_ case as well.

ycombinator commented on util/packages.go:
```diff
@@ -74,8 +74,13 @@ func getPackagePaths(allPaths []string) ([]string, error) {
+			// Not expected file, return nil in order to continue processing sibling directories
+			// Fixes an annoying problem when the .DS_Store file is left behind and the package
+			// is not loading without any error information
+			log.Printf("error: unexpected file: %s", path)
```
Nit: consider changing this to a `warning` instead of an `error` and also mention that we're simply ignoring the unexpected file. That way the person looking at these logs knows this is not a severe problem to worry about.

aleksmaus commented:
Updated the log message and added additional check if dir is not semver.
<img width="897" alt="Screen Shot 2021-01-26 at 8 13 13 AM" src="https://user-images.githubusercontent.com/872351/105849657-ac06b100-5fae-11eb-8bc3-f41c767d35fb.png">

ycombinator reviewed (approved):
LGTM. Thanks for finding and fixing this!

ycombinator commented:
@aleksmaus Apologies for not catching this earlier, but would you mind making a follow up PR to add a CHANGELOG entry for the change in this PR? Thank you.

https://github.com/elastic/package-registry/blob/master/CHANGELOG.md
